### PR TITLE
added pytorch graphconv and tox21 example

### PR DESCRIPTION
**Pull request by evanfeinberg**
In this PR, we introduce an initial implementation of graph convolutions in PyTorch. The implementation can be found in ```contrib/torch/pytorch_graphconv.py```. Briefly summarizing, it requires two matrices as features for each molecule: ```G```, an ```NxN``` adjacency matrix, and ```x```, an ```N x p``` matrix of p features per atom. 

An implementation of both single-task as well as multi-task training is also included. 

We then test this implementation by training a graph convolutional model for toxicity on the Tox21 dataset, achieving a validation AUC of 0.82 with very little hyperparameter tuning. 

Remaining ToDos include implementing a graph pooling layer. 

This work may be particularly helpful for @cxhernandez , Muneeb, and co. 

Thanks to @rbharath for discussions and @miaecle for helpful coding examples.

Edits welcome!!

### CODE REVIEW

**Comment by evanfeinberg** [comment]
The PR also includes a new NxN graph featurizer in feat

**Comment by lilleswing** [comment]
@evanfeinberg sign the CLA https://www.clahub.com/agreements/deepchem/deepchem

File: deepchem/feat/adjacency_fingerprints.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    #atom_matrix = np.zeros((max_n_atoms, n_atom_types+3)).astype(np.uint8)
+    #atom_matrix[:,atom_matrix.shape[1]-1] = 1
+    
+    atom_arrays = []
+    for a_idx in range(0, mol.GetNumAtoms()):
+        atom = mol.GetAtomWithIdx(a_idx)
+        if graph_conv_features:
+            atom_arrays.append(atom_features(atom))
+        else:
+
```
Comment (lilleswing) [inline deepchem/feat/adjacency_fingerprints.py]:
``` python
raise ValueError("Not Implemented Yet")
```
for now?

**Comment by rbharath** [comment]
@evanfeinberg Very cool! I'm glad to see this contributed. Once the CLA is signed, happy to merge this in.

In a future PR, I'd actually encourage merging this into the main DeepChem. I'd love to see more PyTorch support for molecular machine learning :)

**Comment by rbharath** [comment]
Looks like yapf is complaining. Could you run yapf on the files modified within deepchem?
```
pip install yapf
yapf -i deepchem/feat/__init__.py
yapf -i deepchem/feat/adjacency_fingerprints.py
yapf -i deepchem/molnet/load_function/tox21_datasets.py
```

File: deepchem/feat/adjacency_fingerprints.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            atom_matrix[a_idx][-1] = 0
+            atom_matrix[a_idx][atom_type] = 1
+            
+        for n_idx, neighbor in enumerate(atom.GetNeighbors()):
+            if nxn:
+                adj_matrix[a_idx][neighbor.GetIdx()] = 1
+                adj_matrix[a_idx][a_idx] = 1
+            else:
+                adj_matrix[a_idx][n_idx] = neighbor.GetIdx()
+            """
```
Comment (rbharath) [inline deepchem/feat/adjacency_fingerprints.py]: Could you remove the commented out code here?

File: deepchem/molnet/load_function/tox21_datasets.py [hunk 3]
```diff
@@ -37,43 +24,41 @@ def load_tox21(featurizer='ECFP', split='index', reload=True, K=4):
...
-    featurizer = deepchem.feat.WeaveFeaturizer()
+    featurizer = dc.feat.ConvMolFeaturizer()
   elif featurizer == 'Raw':
-    featurizer = deepchem.feat.RawFeaturizer()
+    featurizer = dc.feat.RawFeaturizer()
+  elif featurizer == 'AdjacencyConv':
+    featurizer = dc.feat.AdjacencyFingerprint(max_n_atoms=150, max_valence=6)
 
-  loader = deepchem.data.CSVLoader(
+  loader = dc.data.CSVLoader(
```
Comment (rbharath) [inline deepchem/molnet/load_function/tox21_datasets.py]: We can't use `dc` style imports from within deepchem itself (causes a circular import problem)

File: deepchem/molnet/load_function/tox21_datasets.py [hunk 4]
```diff
@@ -6,29 +6,16 @@
...
 
-def load_tox21(featurizer='ECFP', split='index', reload=True, K=4):
+def load_tox21(featurizer='ECFP', split='index'):
   """Load Tox21 datasets. Does not do train/test split"""
   # Featurize Tox21 dataset
   if "DEEPCHEM_DATA_DIR" in os.environ:
     data_dir = os.environ["DEEPCHEM_DATA_DIR"]
   else:
     data_dir = "/tmp"
-  if reload:
```
Comment (rbharath) [inline deepchem/molnet/load_function/tox21_datasets.py]: Could you add the `reload` logic back in? It's a nice convenience to ensure that datasets are reloaded if they've already been featurized.

File: deepchem/molnet/load_function/tox21_datasets.py [hunk 5]
```diff
@@ -6,29 +6,16 @@
 from __future__ import unicode_literals
 
 import os
-import deepchem
+import deepchem as dc
 
 
-def load_tox21(featurizer='ECFP', split='index', reload=True, K=4):
+def load_tox21(featurizer='ECFP', split='index'):
```
Comment (rbharath) [inline deepchem/molnet/load_function/tox21_datasets.py]: Should restore original fingerprint as well.

File: deepchem/molnet/load_function/tox21_datasets.py [hunk 6]
```diff
@@ -37,43 +24,41 @@ def load_tox21(featurizer='ECFP', split='index', reload=True, K=4):
...
   dataset = loader.featurize(dataset_file, shard_size=8192)
+  #saved_dataset = copy.deepcopy(dataset)
+  #dataset 
 
   # Initialize transformers 
-  transformers = [
-      deepchem.trans.BalancingTransformer(transform_w=True, dataset=dataset)
-  ]
+  transformers = []
+  #transformers = [
```
Comment (rbharath) [inline deepchem/molnet/load_function/tox21_datasets.py]: I'd suggest making this a loading flag to allow users to pick whether or not to do balancing transforms.

File: deepchem/molnet/load_function/tox21_datasets.py [hunk 7]
```diff
@@ -37,43 +24,41 @@ def load_tox21(featurizer='ECFP', split='index', reload=True, K=4):
...
   for transformer in transformers:
     dataset = transformer.transform(dataset)
 
   splitters = {
-      'index': deepchem.splits.IndexSplitter(),
-      'random': deepchem.splits.RandomSplitter(),
-      'scaffold': deepchem.splits.ScaffoldSplitter(),
-      'butina': deepchem.splits.ButinaSplitter(),
-      'task': deepchem.splits.TaskSplitter()
+      'index': dc.splits.IndexSplitter(),
```
Comment (rbharath) [inline deepchem/molnet/load_function/tox21_datasets.py]: s/dc/deepchem/

**Comment by evanfeinberg** [comment]
Thanks for the comments! Will integrate and modify accordingly.

**Comment by lilleswing** [comment]
We should do a cut before adding the pytorch dependency into mainline.

We should probably do a 1.2 cut now.

**Comment by evanfeinberg** [comment]
Thank you @rbharath and @lilleswing for helpful and speedy comments. I believe they have all been integrated now.

**Comment by evanfeinberg** [comment]
A lingering question would be how we would go about merging the PyTorch graphconv code into main...

**Comment by rbharath** [comment]
@evanfeinberg Thanks for the fixes to yapf and the cleanup of graph convs. There are still a few comments outstanding on the changes to `deepchem/molnet/load_function/tox21_datasets.py`. Once those are addressed should be good to merge. It might also be good to just remove those changes from this PR, and add a new loading function that's pytorch specific to `contrib/` rather than modifying the `molnet` loader.

@lilleswing Agreed that we should make a cut before PyTorch support is added to main deepchem. I think we can cut 1.2 after this PR is merged in.

**Comment by evanfeinberg** [comment]
Good catch @rbharath , I think they should all be fixed now. (Had pushed the correct file to the wrong branch initially ;-) )

**Comment by rbharath** [comment]
Thanks, all code changes look good now.

The one last bit is rerunning `yapf` to clear Travis and we should be good:

```
yapf -i deepchem/feat/__init__.py
yapf -i deepchem/feat/adjacency_fingerprints.py
yapf -i deepchem/molnet/load_function/tox21_datasets.py
```

**Comment by evanfeinberg** [comment]
Thanks, done!

**Comment by rbharath** [comment]
LGTM.

**Comment by kexul** [comment]
Hi, @evanfeinberg , I tested the pytorch implement of your work,  your model didn't converge in my experiment. I found that in your implement, the loss function is 
`-(task_label * torch.log(scores)) + (1. - task_label) * torch.log(1. - task_label)`, as far as I know ,the cross entropy loss should be `-log(p) - log(1-p)`. So I modified the second '+' to '-' to achieve converge in my expirement. Another question is that in your ipython notebook, the RMSE valuse was keep rising along training, but the mean roc auc value is raising at the same time, how can this be possible? Isn't the RMSE value the lower, better the predition is? 
Any help is appreciated !
